File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        itr = head

        while itr:
            length += 1
            itr = itr.next
        
        itr = head
        prev = None
        for i in range(length - n):
            prev = itr
            itr = itr.next

        if not prev:
            next = itr.next
            itr.next = None
            return next
        
        prev.next = itr.next
        itr.next = None
        return head

File: Data Structures & Algorithms/maximum-twin-sum-of-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        vector<int> vals;

        for (ListNode* curr{head}; curr != nullptr; curr = curr->next) {
            vals.push_back(curr->val);
        }

        int l{}, r{static_cast<int>(vals.size()) - 1}, res{};
        while (l < r) {
            res = max(res, vals[l++] + vals[r--]);
        }

        return res;
    }
};